lib/ecmascript.ts

```diff
@@ -836,17 +883,350 @@ export const ES = ObjectAssign({}, ES2020, {
       millisecond,
       microsecond,
       nanosecond,
-      calendar
+      offsetBehaviour,
+      offsetNs,
+      timeZone,
+      'compatible',
+      'reject'
     );
-  },
-  ValidateTemporalUnitRange: (largestUnit, smallestUnit) => {
-    if (ALLOWED_UNITS.indexOf(largestUnit) > ALLOWED_UNITS.indexOf(smallestUnit)) {
-      throw new RangeError(`largestUnit ${largestUnit} cannot be smaller than smallestUnit ${smallestUnit}`);
-    }
-  },
-  DefaultTemporalLargestUnit: (
-    years,
-    months,
+    return CreateTemporalZonedDateTime(epochNanoseconds, timeZone, calendar);
+  }
+  return CreateTemporalDateTime(year, month, day, hour, minute, second, millisecond, microsecond, nanosecond, calendar);
+}
+
+export function ValidateTemporalUnitRange(largestUnit, smallestUnit) {
+  if (ALLOWED_UNITS.indexOf(largestUnit) > ALLOWED_UNITS.indexOf(smallestUnit)) {
+    throw new RangeError(`largestUnit ${largestUnit} cannot be smaller than smallestUnit ${smallestUnit}`);
+  }
+}
+
+export function DefaultTemporalLargestUnit(
+  years,
+  months,
+  weeks,
+  days,
+  hours,
+  minutes,
+  seconds,
+  milliseconds,
+  microseconds,
+  nanoseconds
+) {
+  const singular = new Map(SINGULAR_PLURAL_UNITS);
+  for (const [prop, v] of [
+    ['years', years],
+    ['months', months],
+    ['weeks', weeks],
+    ['days', days],
+    ['hours', hours],
+    ['minutes', minutes],
+    ['seconds', seconds],
+    ['milliseconds', milliseconds],
+    ['microseconds', microseconds],
+    ['nanoseconds', nanoseconds]
+  ]) {
+    if (v !== 0) return singular.get(prop);
+  }
+  return 'nanosecond';
+}
+
+export function LargerOfTwoTemporalUnits(unit1, unit2) {
+  if (ALLOWED_UNITS.indexOf(unit1) > ALLOWED_UNITS.indexOf(unit2)) return unit2;
+  return unit1;
+}
+
+export function CastIfDefined(value, cast) {
+  if (value !== undefined) {
+    return cast(value);
+  }
+  return value;
+}
+
+export function ToPartialRecord(bag, fields, callerCast?: (value: unknown) => unknown) {
+  if (Type(bag) !== 'Object') return undefined;
+  let any;
+  for (const property of fields) {
+    const value = bag[property];
+    if (value !== undefined) {
+      any = any || {};
+      if (callerCast === undefined && BUILTIN_CASTS.has(property)) {
+        any[property] = BUILTIN_CASTS.get(property)(value);
+      } else if (callerCast !== undefined) {
+        any[property] = callerCast(value);
+      } else {
+        any[property] = value;
+      }
+    }
+  }
+  return any ? any : undefined;
+}
+
+interface TemporalFields {
```

Comment:
This is very similar to the publicly-exposed type `Temporal.ZonedDateTimeLike`. The only difference is lack of `calendar` and the type of the `timeZone` property.  Should we derive this type from the public type or leave it as-is here?  I don't have an opinion either way, but seems like we should either be consistently re-using public types where we can, or should never be re-using them.

Reply:
I think there might be a way to make this use the TS Mapped Types functionality to be able to type it based on the fields you give it. Will investigate in a follow-up PR.

I'm slightly hesitant to re-use the public API types given that those might change and while that might be right for the external API it could easily become incorrect for part of the internal API.